crates/store/src/filesystem/mod.rs

```diff
@@ -0,0 +1,412 @@
+        let bin = self
+            .read_bytes(&bin_link_file)
+            .await
+            .map_err(|e| Error::BadBinFile(identity, Box::new(e)).into_get_error())?;
+        Ok((meta, Cursor::new(bin)))
+    }
+}
+
+#[async_trait]
+impl<K> Keys<K> for Filesystem<K>
+where
+    K: DeserializeOwned + 'static + Sync + Send + Clone,
+{
+    type Stream = Iter<std::vec::IntoIter<Result<K, Error>>>;
+    type StreamError = Error;
+
+    async fn keys(&self) -> Self::Stream {
+        iter(self.get_keys().await)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::{
+        fs::{create_dir, remove_dir_all, File},
+        path::Path,
+    };
+
+    use futures::{AsyncReadExt, AsyncWriteExt, StreamExt};
+
+    const TEST_DATA_DIRECTORY: &str = ".test_data";
+
+    #[tokio::test]
+    async fn test() {
+        let key = "test".to_string();
+        let mime = "text/plain".parse::<Mime>().unwrap();
+        let path = Path::new(TEST_DATA_DIRECTORY);
+
+        let _ = remove_dir_all(path);
+        create_dir(path).unwrap();
+        let root_dir = File::open(path).unwrap();
+
+        let mut fs = Filesystem::new(Dir::from_std_file(root_dir.into()), "somestore".to_string());
+
+        assert_eq!(
+            format!(
+                "{:?}",
+                fs.get(key.clone())
+                    .await
+                    .map(|(x, y)| (x, y.into_inner()))
+                    .unwrap_err()
+            ),
+            format!("{:?}", GetError::NotFound::<Error>)
+        );
+
+        assert!(fs.keys().await.collect::<Vec<_>>().await.is_empty());
+
+        let hash = {
+            let mut hasher = Algorithms::default().writer(io::sink());
+            assert!(matches!(hasher.write_all(&[42]).await, Ok(())));
+            hasher.digests()
+        };
+
+        {
+            let w = fs.create(key.clone(), mime.clone()).await;
+            assert!(matches!(w, Ok(_)));
+            let mut w = w.unwrap();
+            assert!(matches!(w.write_all(&[42]).await, Ok(())));
+            assert_eq!(w.finish().await, (1, hash.clone()))
+        }
+
+        {
+            let mr = fs.get(key.clone()).await;
+            assert!(matches!(mr, Ok(_)));
+            let mut v = vec![];
+            let (meta, mut r) = mr.unwrap();
+            assert_eq!(
+                meta,
+                Meta {
+                    mime: mime.clone(),
+                    size: 1,
+                    hash,
+                }
+            );
+            assert!(matches!(r.read_to_end(&mut v).await, Ok(1)))
+        }
+
+        assert!(matches!(
+            fs.create(key.clone(), mime).await,
+            Err(CreateError::Occupied)
+        ));
+
+        assert_eq!(
+            fs.keys()
+                .await
+                .collect::<Vec<_>>()
+                .await
+                .into_iter()
+                .map(|k| k.unwrap())
+                .collect::<Vec<String>>(),
```

Comment:
What is that first collect that returns a future?

Reply:
`.keys` the keys trait


```rust
#[async_trait]
pub trait Keys<K>
where
    K: DeserializeOwned + 'static,
{
    type Stream: Send + Stream<Item = Result<K, Self::StreamError>>;
    type StreamError: Sync + Send + std::error::Error;

    async fn keys(&self) -> Self::Stream;
}
```